File: YourPastaFE/src/app/service/inventario.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Inventario } from '../dto/inventario.model';

@Injectable({
  providedIn: 'root'
})
export class InventarioService {
  private readonly API = 'http://localhost:8080/api/inventario';

  constructor(private http: HttpClient) {}

  // Crea nuovo articolo
  // Prende un oggetto articolo dal frontend, manda una richiesta POST e riceve l'articolo creato dal backend
  creaArticolo(articolo: Inventario): Observable<Inventario> {
    return this.http.post<Inventario>(this.API, articolo);
  }

  // Trova tutti gli articoli
  // Non prende nulla dal frontend, manda una richiesta GET e riceve una lista di articoli dal backend
  trovaTuttiGliArticoli(): Observable<Inventario[]> {
    return this.http.get<Inventario[]>(this.API);
  }

  // Trova articolo per ID
  // Prende l'id dal frontend, manda una richiesta GET e riceve l'articolo corrispondente dal backend
  trovaArticoloPerId(id: number): Observable<Inventario> {
    return this.http.get<Inventario>(`${this.API}/${id}`);
  }

  // Trova articoli per categoria
  // Prende la categoria dal frontend, manda una richiesta GET e riceve una lista di articoli filtrati dal backend
  trovaArticoliPerCategoria(categoria: string): Observable<Inventario[]> {
    return this.http.get<Inventario[]>(`${this.API}/categoria/${categoria}`);
  }

  // Cerca articoli per nome
  // Prende il nome dal frontend, manda una richiesta GET con query param e riceve una lista di articoli dal backend
  cercaArticoliPerNome(nome: string): Observable<Inventario[]> {
    return this.http.get<Inventario[]>(`${this.API}/cerca?nome=${nome}`);
  }

  // Aggiorna quantità articolo
  // Prende id e quantità dal frontend, manda una richiesta PUT e riceve l'articolo aggiornato dal backend
  aggiornaQuantita(id: number, quantita: number): Observable<Inventario> {
    return this.http.put<Inventario>(`${this.API}/${id}/quantita`, { quantita });
  }

  // Aggiorna prezzo articolo
  // Prende id e prezzo dal frontend, manda una richiesta PUT e riceve l'articolo aggiornato dal backend
  aggiornaPrezzo(id: number, prezzo: number): Observable<Inventario> {
    return this.http.put<Inventario>(`${this.API}/${id}/prezzo`, { prezzo });
  }

  // Aggiorna inventario completo
  // Prende id e oggetto inventario dal frontend, manda una richiesta PUT e riceve l'inventario aggiornato dal backend
  aggiornaInventario(id: number, inventario: Inventario): Observable<Inventario> {
    return this.http.put<Inventario>(`${this.API}/${id}`, inventario);
  }

  // Elimina articolo
  // Prende l'id dal frontend, manda una richiesta DELETE e riceve conferma di eliminazione dal backend
  eliminaArticolo(id: number): Observable<any> {
    return this.http.delete(`${this.API}/${id}`);
  }
}

File: YourPastaBE/src/main/java/it/unife/ingsw20242025/YourPasta/service/InventarioService.java
package it.unife.ingsw20242025.YourPasta.service;

import it.unife.ingsw20242025.YourPasta.model.Inventario;
import it.unife.ingsw20242025.YourPasta.repository.InventarioRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service // Servizio che gestisce la logica di business per l'inventario

public class InventarioService {
    
    // Dipendenze: repository e servizio contatore, iniettate tramite costruttore
    private final InventarioRepository inventarioRepository;
    private final ContatoreService contatoreService; // Per generare gli ID
    
    public InventarioService(InventarioRepository inventarioRepository, ContatoreService contatoreService) {
        this.inventarioRepository = inventarioRepository;
        this.contatoreService = contatoreService;
    }
    
    // Crea un nuovo articolo in inventario
    // non lo chiamo ingrediente perche ci sono anche le bevande (non so se sia stata una furbata in effetti)
    public Inventario creaArticolo(String nome, Inventario.Categoria categoria, Integer quantita, Double prezzoUnitario) {
        
        // Genera un nuovo ID usando il contatore
        Integer nuovoId = contatoreService.getNextId("inventario");
        
        // Crea l'articolo
        Inventario articolo = new Inventario();
        articolo.setId(nuovoId);
        articolo.setNome(nome);
        articolo.setCategoria(categoria);
        articolo.setQuantita(quantita);
        articolo.setPrezzoUnitario(prezzoUnitario);
        
        return inventarioRepository.save(articolo);
    }
    
    // Trova articolo per ID
    public Optional<Inventario> trovaArticoloPerId(Integer id) {
        return inventarioRepository.findById(id);
    }
    
    // Trova tutti gli articoli
    public List<Inventario> trovaTuttiGliArticoli() {
        return inventarioRepository.findAll();
    }
    
    // Trova articoli per categoria
    public List<Inventario> trovaArticoliPerCategoria(Inventario.Categoria categoria) {
        return inventarioRepository.findByCategoria(categoria);
    }
    
    // Trova articoli per nome (ricerca parziale)
    public List<Inventario> trovaArticoliPerNome(String nome) {
        return inventarioRepository.findByNomeContainingIgnoreCase(nome);
    }
    
    // Aggiorna quantità di un articolo
    public Inventario aggiornaQuantita(Integer id, Integer nuovaQuantita) {
        Inventario articolo = inventarioRepository.findById(id).orElseThrow(() -> new RuntimeException("Articolo non trovato con ID: " + id));
        
        articolo.setQuantita(nuovaQuantita);
        return inventarioRepository.save(articolo);
    }
    
    // Aggiorna prezzo di un articolo
    public Inventario aggiornaPrezzo(Integer id, Double nuovoPrezzo) {
        Inventario articolo = inventarioRepository.findById(id).orElseThrow(() -> new RuntimeException("Articolo non trovato con ID: " + id));
        
        articolo.setPrezzoUnitario(nuovoPrezzo);
        return inventarioRepository.save(articolo);
    }
    

    // Aggiorna tutti i campi di un articolo
    public Inventario aggiornaArticolo(Integer id, String nome, Inventario.Categoria categoria, Integer quantita, Double prezzoUnitario) {
        Inventario articolo = inventarioRepository.findById(id).orElseThrow(() -> new RuntimeException("Articolo non trovato con ID: " + id));

        articolo.setNome(nome);
        articolo.setCategoria(categoria);
        articolo.setQuantita(quantita);
        articolo.setPrezzoUnitario(prezzoUnitario);
        return inventarioRepository.save(articolo);
    }

    // Elimina articolo
    public void eliminaArticolo(Integer id) {
        if (!inventarioRepository.existsById(id)) {
            throw new RuntimeException("Articolo non trovato con ID: " + id);
        }
        inventarioRepository.deleteById(id);
    }
}